quinn-udp/src/unix.rs

```diff
@@ -66,6 +66,100 @@ impl Default for UdpSocketState {
     }
 }
 
+/// Fallback implementation of `sendmmsg` using `sendmsg`
+/// for systems which do not support `sendmmsg`
+/// such as Linux <3.0.
+#[cfg(not(any(target_os = "macos", target_os = "ios")))]
+unsafe fn sendmmsg_fallback(
+    sockfd: libc::c_int,
+    msgvec: *mut libc::mmsghdr,
+    vlen: libc::c_uint,
+) -> libc::c_int {
+    let flags = 0;
+    if vlen == 0 {
+        return 0;
+    }
+
+    let n = libc::sendmsg(sockfd, &(*msgvec).msg_hdr, flags);
+    if n == -1 {
+        -1
+    } else {
+        (*msgvec).msg_len = n as libc::c_uint;
+        1
+    }
+}
+
+#[cfg(not(any(target_os = "macos", target_os = "ios")))]
+unsafe fn sendmmsg_with_fallback(
+    sockfd: libc::c_int,
+    msgvec: *mut libc::mmsghdr,
+    vlen: libc::c_uint,
+) -> libc::c_int {
+    let flags = 0;
+    let ret = libc::syscall(libc::SYS_sendmmsg, sockfd, msgvec, vlen, flags) as libc::c_int;
+
+    if ret == -1 {
+        let e = io::Error::last_os_error();
+        match e.raw_os_error() {
+            Some(libc::ENOSYS) => {
+                // Fallback to `sendmsg`.
+                sendmmsg_fallback(sockfd, msgvec, vlen)
+            }
+            _ => -1,
+        }
+    } else {
+        ret
+    }
+}
+
+/// Fallback implementation of `recvmmsg` using `recvmsg`
+/// for systems which do not support `recvmmsg`
+/// such as Linux <2.6.33.
+#[cfg(not(any(target_os = "macos", target_os = "ios")))]
+unsafe fn recvmmsg_fallback(
+    sockfd: libc::c_int,
+    msgvec: *mut libc::mmsghdr,
+    vlen: libc::c_uint,
+) -> libc::c_int {
+    let flags = 0;
+    if vlen == 0 {
+        return 0;
+    }
+
+    let n = libc::recvmsg(sockfd, &mut (*msgvec).msg_hdr, flags);
+    if n == -1 {
+        -1
+    } else {
+        (*msgvec).msg_len = n as libc::c_uint;
+        1
+    }
+}
+
+#[cfg(not(any(target_os = "macos", target_os = "ios")))]
+unsafe fn recvmmsg_with_fallback(
+    sockfd: libc::c_int,
+    msgvec: *mut libc::mmsghdr,
+    vlen: libc::c_uint,
+) -> libc::c_int {
+    let flags = 0;
+    let timeout = ptr::null_mut::<libc::timespec>();
+    let ret =
+        libc::syscall(libc::SYS_recvmmsg, sockfd, msgvec, vlen, flags, timeout) as libc::c_int;
+
+    if ret == -1 {
+        let e = io::Error::last_os_error();
+        match e.raw_os_error() {
+            Some(libc::ENOSYS) => {
+                // Fallback to `recvmsg`.
+                recvmmsg_fallback(sockfd, msgvec, vlen)
```

Comment:
Don't know how much you care, but presumably we *could* store state on whether mmsg support is available somewhere, maybe in the `UdpState`, so that we don't have to fallback everytime? (I suppose that would come at the cost of one extra branch even for environments where mmsg support *is* available, but that should be a pretty cheap branch with prediction.)